File: src/main/java/org/spartandevs/customdeathmessages/util/DiscordManager.java
package org.spartandevs.customdeathmessages.util;

import net.essentialsx.api.v2.services.discord.DiscordService;
import net.essentialsx.api.v2.services.discord.MessageType;
import org.bukkit.entity.Player;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.spartandevs.customdeathmessages.CustomDeathMessages;

public class DiscordManager {
    private final DiscordSender discord;

    public DiscordManager(CustomDeathMessages plugin) {
        DiscordSender sender = (message, player, event) -> {
        };

        if (plugin.getServer().getPluginManager().isPluginEnabled("EssentialsDiscord")) {
            DiscordService essentialsDiscord = plugin.getServer().getServicesManager().load(DiscordService.class);

            if (essentialsDiscord != null) {
                sender = (message, player, event) ->
                        essentialsDiscord.sendMessage(MessageType.DefaultTypes.DEATH, ":skull: " + message, false);

                plugin.getLogger().info("Successfully loaded EssentialsDiscord support.");
            }
        } else if (plugin.getServer().getPluginManager().isPluginEnabled("DiscordSRV")) {
            plugin.getLogger().info("DiscordSRV support is not yet implemented. EssentialsDiscord is currently the only supported Discord plugin. This will be fixed in a future update.");
            // sender = (message, player, event) -> {
            //     SRVSender.send(message, player, event);
            //     plugin.getLogger().info("Successfully loaded DiscordSRV support.");
            // };
        }

        this.discord = sender;
    }

    public void sendDeathMessage(String message, Player player, PlayerDeathEvent event) {
        discord.send(message, player, event);
    }

    @FunctionalInterface
    private interface DiscordSender {
        void send(String message, Player player, PlayerDeathEvent event);
    }
}

File: src/main/java/org/spartandevs/customdeathmessages/util/DiscordManager.kt
package org.spartandevs.customdeathmessages.util

import net.essentialsx.api.v2.services.discord.DiscordService
import net.essentialsx.api.v2.services.discord.MessageType
import org.bukkit.entity.Player
import org.bukkit.event.entity.PlayerDeathEvent
import org.spartandevs.customdeathmessages.CustomDeathMessages


class DiscordManager(plugin: CustomDeathMessages) {
    private var discord: (message: String, player: Player, event: PlayerDeathEvent) -> Unit = { _, _, _ -> }

    init {
        if (plugin.server.pluginManager.isPluginEnabled("EssentialsDiscord")) {
            val essDiscord = plugin.server.servicesManager.load(DiscordService::class.java)

            essDiscord?.let {
                discord = { message, _, _ ->
                    it.sendMessage(MessageType.DefaultTypes.DEATH, ":skull: $message", false)
                }

                plugin.logger.info("Successfully loaded EssentialsDiscord support.")
            }
        } else if (plugin.server.pluginManager.isPluginEnabled("DiscordSRV")) {
            discord = { message, player, event ->
                SRVSender.send(message, player, event);
                plugin.logger.info("Successfully loaded DiscordSRV support.")
            }
        }
    }

    fun sendDeathMessage(message: String, player: Player, event: PlayerDeathEvent) {
        discord(message, player, event)
    }
}
